lisa-kernel/src/main/scala/lisa/kernel/fol/EquivalenceChecker.scala

```diff
@@ -14,88 +14,427 @@ import scala.math.Numeric.IntIsIntegral
+  }
+  def isSameSet(s1: Set[Formula], s2: Set[Formula]): Boolean =
+    isSubset(s1, s2) && isSubset(s2, s1)
+
+  def contains(s: Set[Formula], f: Formula): Boolean = {
+    s.exists(g => isSame(f, g))
+  }
+
+
+
+
+
+
+  private var totPolarFormula = 0
   sealed abstract class SimpleFormula {
+    val uniqueKey: Int = totPolarFormula
+    totPolarFormula += 1
     val size: Int
+    var inverse: Option[SimpleFormula] = None
     private[EquivalenceChecker] var normalForm: Option[NormalFormula] = None
   }
-  case class SimplePredicate(id: PredicateLabel, args: List[Term]) extends SimpleFormula {
+  case class SimplePredicate(id: PredicateLabel, args: Seq[Term], polarity:Boolean, original:Formula) extends SimpleFormula {
+    override def toString:String = s"SimplePredicate($id, $args, $polarity)"
     val size = 1
   }
-  case class SimpleConnector(id: ConnectorLabel, args: List[SimpleFormula]) extends SimpleFormula {
+  case class SimpleSchemConnector(id: SchematicConnectorLabel, args: Seq[SimpleFormula], polarity:Boolean) extends SimpleFormula {
     val size = 1
   }
-  case class SNeg(child: SimpleFormula) extends SimpleFormula {
-    val size: Int = 1 + child.size
-  }
-  case class SOr(children: List[SimpleFormula]) extends SimpleFormula {
+  case class SimpleAnd(children: Seq[SimpleFormula], polarity:Boolean) extends SimpleFormula {
     val size: Int = (children map (_.size)).foldLeft(1) { case (a, b) => a + b }
   }
-  case class SForall(x: Identifier, inner: SimpleFormula) extends SimpleFormula {
+  case class SimpleForall(x: Identifier, inner: SimpleFormula, polarity:Boolean) extends SimpleFormula {
     val size: Int = 1 + inner.size
   }
-  case class SExists(x: Identifier, inner: SimpleFormula) extends SimpleFormula {
-    val size: Int = 1 + inner.size
-  }
-  case class SLiteral(b: Boolean) extends SimpleFormula {
+  case class SimpleLiteral(polarity: Boolean) extends SimpleFormula {
     val size = 1
-    normalForm = Some(NLiteral(b))
+    normalForm = Some(NormalLiteral(polarity))
   }
+
+
+
+  private var totNormalFormula = 0
   sealed abstract class NormalFormula {
-    val code: Int
+    val uniqueKey: Int = totNormalFormula
+    totNormalFormula += 1
+    //val code: Int
+    var formulaP: Option[Formula] = None
+    var formulaN: Option[Formula] = None
+    var formulaAIG: Option[Formula] = None
+    var inverse: Option[NormalFormula] = None
+
+    private val lessThanBitSet: mutable.Set[Long] = new mutable.HashSet()
+    setLessThanCache(this, true)
+
+    def lessThanCached(other: NormalFormula): Option[Boolean] = {
+      val otherIx = 2 * other.uniqueKey
+      if (lessThanBitSet.contains(otherIx)) Some(lessThanBitSet.contains(otherIx + 1))
+      else None
+    }
+
+    def setLessThanCache(other: NormalFormula, value: Boolean): Unit = {
+      val otherIx = 2 * other.uniqueKey
+      lessThanBitSet.contains(otherIx)
+      if (value) lessThanBitSet.update(otherIx + 1, true)
+    }
+
+    //TODO: Check if needed
```

Comment:
What does this do exactly? Any -> Boolean calling super.equals feels scary to have. I could not find any references to it in the rest of the code either.